lib/msf/core/windows_version.rb

```diff
@@ -0,0 +1,160 @@
+# -*- coding: binary -*-
+module Msf
+  #
+  # Represents the version of a Windows operating system
+  #
+  class WindowsVersion
+
+    VER_NT_WORKSTATION = 1
+    VER_NT_DOMAIN_CONTROLLER = 2
+    VER_NT_SERVER = 3
+
+    XP_SP0 = Rex::Version.new('5.1.2600.0')
+    XP_SP1 = Rex::Version.new('5.1.2600.1')
+    XP_SP2 = Rex::Version.new('5.1.2600.2')
+    XP_SP3 = Rex::Version.new('5.1.2600.3')
+    Server2003_SP0 = Rex::Version.new('5.2.3790.0')
+    Server2003_SP1 = Rex::Version.new('5.2.3790.1')
+    Server2003_SP2 = Rex::Version.new('5.2.3790.2')
+    Vista_SP0 = Server2008_SP0 = Rex::Version.new('6.0.6000.0')
+    Vista_SP1 = Server2008_SP1 = Rex::Version.new('6.0.6001.1')
+    Vista_SP2 = Server2008_SP2 = Rex::Version.new('6.0.6002.2')
+    Win7_SP0 = Server2008_R2_SP0 = Rex::Version.new('6.1.7600.0')
+    Win7_SP1 = Server2008_R2_SP1 = Rex::Version.new('6.1.7601.1')
+    Win8 = Server2012 = Rex::Version.new('6.2.9200.0')
+    Win81 = Server2012_R2 = Rex::Version.new('6.3.9600.0')
+    Win10_1507 = Win10_InitialRelease = Rex::Version.new('10.0.10240.0')
+    Win10_1511 = Rex::Version.new('10.0.10586.0')
+    Win10_1607 = Server2016 = Rex::Version.new('10.0.14393.0')
+    Win10_1703 = Rex::Version.new('10.0.15063.0')
+    Win10_1709 = Rex::Version.new('10.0.16299.0')
+    Win10_1803 = Rex::Version.new('10.0.17134.0')
+    Win10_1809 = Server2019 = Rex::Version.new('10.0.17763.0')
+    Win10_1903 = Rex::Version.new('10.0.18362.0')
+    Win10_1909 = Rex::Version.new('10.0.18363.0')
+    Win10_2004 = Rex::Version.new('10.0.19041.0')
+    Win10_20H2 = Rex::Version.new('10.0.19042.0')
+    Win10_21H1 = Rex::Version.new('10.0.19043.0')
+    Win10_21H2 = Rex::Version.new('10.0.19044.0')
+    Win10_22H2 = Rex::Version.new('10.0.19044.0')
+    Server2022 = Rex::Version.new('10.0.20348.0')
+    Win11_21H2 = Rex::Version.new('10.0.22000.0')
+    Win11_22H2 = Rex::Version.new('10.0.22621.0')
+
+    module MajorRelease
+      NT351 = "Windows NT 3.51"
+      Win95 = "Windows 95"
+      Win98 = "Windows 98"
+      WinME = "Windows ME"
+
+      XP = "Windows XP"
+      Server2003 = "Windows Server 2003"
+
+      Vista = "Windows Vista"
+      Server2008 = "Windows Server 2008"
+      
+      Win7 = "Windows 7"
+      Server2008R2 = "Windows 2008 R2"
+
+      Win8 = "Windows 8"
+      Server2012 = "Windows Server 2012"
+
+      Win81 = "Windows 8.1"
+      Server2012R2 = "Windows Server 2012 R2"
+
+      Win10Plus = "Windows 10+"
```

Comment:
On one hand this makes the most intuitive sense... on the other, the way Microsoft actually treats their version numbers makes this a less obvious choice in my opinion: Windows 10 is 10.0.x whereas Windows 11 is... also 10.0.x. Combine that with the versioning of Windows Server lately, wherein Server 2016 is effectively in line with Windows 10 v1607 (in terms of a shared build number), Server 2019 with Win10 1809 and Server 2022... well it's no longer in line with a specific workstation version, in terms of its build number... but they're all still "Windows 10.0.x". It seems to me that, if we ignore the marketing of what they call it, we're kind of in an era of incremental releases, rather than the former monolithic quantum leap. 

The other consideration I had is around forwards-compatibility, which is what this PR aims to resolve. It's clear that Microsoft are changing their version numbering strategy, where everything is now 10.0.x; but it's not clear to me what it will be in the future, and so I can't say with confidence what a check for "Windows 11 in general" would actually look like? Would it check for "10.0.x where x is greater than 22000"? Could that misidentify something as Windows 12 in the future? I know playing the what-if game is maybe a little fraught, and hey, you can always change code in the future... but in the past, it was easy: a new "Major Release" means a new minor version number at least (Vista == 6.0, Win7 == 6.1, Win8 == 6.2, Win8.1 == 6.3). Now it seems much less clear.

I guess tl;dr I'm keen to see the dust settle a bit more on the version numbers before confidently proclaiming "This is Windows 11".